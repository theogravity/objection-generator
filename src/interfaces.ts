export interface YamlConfig {
  config: {
    model: {
      classNamePrefix?: string
      classNamePostfix?: string
    }
  }
  models: {
    [key: string]: ModelDef
  }
  components: {
    [key: string]: any
  }
}

export interface ModelDef {
  tableName: string
  jsonSchema: {
    required?: string[]
    properties: {
      [key: string]: {
        type: string
        [key: string]: any
      }
    }
  }
}

export interface Arguments {
  specFile: string
  outDir: string
  templateDir?: string
}
