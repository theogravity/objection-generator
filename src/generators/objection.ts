import Handlebars from 'handlebars'
import { promises } from 'fs'
import fsEx from 'fs-extra'
import { BaseGenerator } from './base'

export class ObjectionGenerator extends BaseGenerator {
  async generate () {
    await fsEx.ensureDir(this.outDir + '/models')
    await this.generateModels()

    fsEx.copy(
      `${this.templatesDir}/BaseModel.ts`,
      `${this.outDir}/models/BaseModel.ts`
    )
  }

  protected async generateModels () {
    for (const file of this.templateFiles) {
      if (file.includes('$$model$$')) {
        const hbsData = await promises.readFile(file, 'utf8')
        const template = Handlebars.compile(hbsData)

        for (const result of this.compileModel(template)) {
          await promises.writeFile(
            `${this.outDir}/models/${result.modelName}.ts`,
            this.prettify(result.content),
            'utf8'
          )
        }
      }
    }
  }

  protected * compileModel (template: HandlebarsTemplateDelegate) {
    for (const model of Object.keys(this.defs.models)) {
      const prefix = this.defs.config?.model?.classNamePrefix || ''
      const postfix = this.defs.config?.model?.classNamePostfix || ''
      const modelName = `${prefix}${model}${postfix}`

      const modelData = this.defs.models[model]

      const templateData = {
        modelPrefix: prefix,
        modelPostfix: postfix,
        modelName,
        data: modelData
      }

      yield {
        modelName,
        content: template(templateData)
      }
    }
  }
}
