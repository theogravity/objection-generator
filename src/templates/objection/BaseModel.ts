import { Model } from 'objection'

export class BaseModel extends Model {}
