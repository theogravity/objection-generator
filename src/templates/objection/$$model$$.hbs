{{#if data.relations}}
import { Model } from 'objection'
import path from 'path'
{{/if}}

import { BaseModel } from './BaseModel'

{{#each data.jsonSchema.properties}}
  {{#if enum}}
export enum {{{@root.modelNofix}}}{{{firstCap (camelCase @key)}}}Enum {
  {{{defineEnum .}}}
}

  {{/if}}
{{/each}}

export class {{{modelName}}} extends BaseModel {
{{#each data.jsonSchema.properties}}
  {{#if enum}}
  {{{@key}}}: {{{@root.modelNofix}}}{{{firstCap (camelCase @key)}}}Enum
  {{else}}
  {{{@key}}}: {{{tsType .}}}
  {{/if}}
{{/each}}

  static tableName = '{{{data.tableName}}}'

  static get jsonSchema() {
    return {{{printSchema data.jsonSchema}}}
  }

  {{#if data.idColumn}}
  static get idColumn() {
    return '{{{data.idColumn}}}'
  }
  {{/if}}

  {{#if data.relations}}
  static get relationMappings () {
    return {
      {{#each data.relations}}
        {{@key}}: {
        relation: {{{relation}}},
        modelClass: path.join(__dirname, '{{{getModelName @root.modelPrefix @root.modelPostfix modelClass}}}'),
        {{{formatRelation .}}}
        },
      {{/each}}
    }
  }
  {{/if}}
}
