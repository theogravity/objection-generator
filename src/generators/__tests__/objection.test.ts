/* eslint-env jest */

import { pathExistsSync, remove } from 'fs-extra'
import { readFileSync } from 'fs'
import { resolve } from 'path'
import validConfig from '../../__fixtures__/valid-config.json'

import { ObjectionGenerator } from '../objection'
import { YamlConfig } from '../../interfaces'

const templateDir = resolve(__dirname, '..', '..', 'templates')

const OUT_DIR = '/tmp/objection-generator-test/generator/objection'

beforeEach(async () => {
  await remove(OUT_DIR)
})

describe('ObjectionGenerator class', () => {
  it('should generate models', async () => {
    const generator = new ObjectionGenerator(
      (validConfig as unknown) as YamlConfig,
      templateDir + '/objection',
      OUT_DIR
    )

    await generator.init()
    await generator.generate()

    const movieModelPath = resolve(OUT_DIR, 'models', 'MovieModel.ts')

    expect(pathExistsSync(movieModelPath)).toBe(true)
    expect(readFileSync(movieModelPath, 'utf8')).toMatchSnapshot()
  })
})
