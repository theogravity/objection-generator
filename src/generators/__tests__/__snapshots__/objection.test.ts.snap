// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectionGenerator class should generate models 1`] = `
"import path from 'path'

import { BaseModel } from './BaseModel'

export class MovieModel extends BaseModel {
  id: string
  name: string

  static tableName = 'movies'

  static get jsonSchema () {
    return {
      type: 'object',
      required: ['name'],
      properties: {
        id: {
          type: 'string'
        },
        name: {
          type: 'string',
          minLength: 1,
          maxLength: 255
        }
      }
    }
  }

  static get relationMappings () {
    return {
      reviews: {
        relation: Model.HasManyRelation,
        modelClass: path.join(__dirname, 'ReviewModel'),
        join: {
          from: 'movie.id',
          to: 'review.movieId'
        }
      }
    }
  }
}
"
`;
